docs/client-libraries-consumers.md

```diff
@@ -0,0 +1,935 @@
+
+````mdx-code-block
+<Tabs groupId="lang-choice"
+  defaultValue="C++"
+  values={[{"label":"C++","value":"C++"},{"label":"Go","value":"Go"}]}>
+<TabItem value="C++">
+
+This example starts a subscription at the earliest offset and consumes 100 messages.
+
+```cpp
+#include <pulsar/Client.h>
+#include <atomic>
+#include <thread>
+
+using namespace pulsar;
+
+std::atomic<uint32_t> messagesReceived;
+
+void handleAckComplete(Result res) {
+    std::cout << "Ack res: " << res << std::endl;
+}
+
+void listener(Consumer consumer, const Message& msg) {
+    std::cout << "Got message " << msg << " with content '" << msg.getDataAsString() << "'" << std::endl;
+    messagesReceived++;
+    consumer.acknowledgeAsync(msg.getMessageId(), handleAckComplete);
+}
+
+int main() {
+    Client client("pulsar://localhost:6650");
+
+    Consumer consumer;
+    ConsumerConfiguration config;
+    config.setMessageListener(listener);
+    config.setSubscriptionInitialPosition(InitialPositionEarliest);
+    Result result = client.subscribe("persistent://public/default/my-topic", "consumer-1", config, consumer);
+    if (result != ResultOk) {
+        std::cout << "Failed to subscribe: " << result << std::endl;
+        return -1;
+    }
+
+    // wait for 100 messages to be consumed
+    while (messagesReceived < 100) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
+
+    std::cout << "Finished consuming asynchronously!" << std::endl;
+
+    client.close();
+    return 0;
+}
+```
+
+</TabItem>
+<TabItem value="Go">
+
+```go
+import (
+    "fmt"
+    "log"
+
+    "github.com/apache/pulsar-client-go/pulsar"
+)
+
+func main() {
+    client, err := pulsar.NewClient(pulsar.ClientOptions{URL: "pulsar://localhost:6650"})
+    if err != nil {
+        log.Fatal(err)
+    }
+
+    defer client.Close()
+
+    // we can listen this channel
+    channel := make(chan pulsar.ConsumerMessage, 100)
+
+    options := pulsar.ConsumerOptions{
+        Topic:            "topic-1",
+        SubscriptionName: "my-subscription",
+        Type:             pulsar.Shared,
+        // fill `MessageChannel` field will create a listener
+        MessageChannel: channel,
+    }
+
+    consumer, err := client.Subscribe(options)
+    if err != nil {
+        log.Fatal(err)
+    }
+
+    defer consumer.Close()
+
+    // Receive messages from channel. The channel returns a struct `ConsumerMessage` which contains message and the consumer from where
+    // the message was received. It's not necessary here since we have 1 single consumer, but the channel could be
+    // shared across multiple consumers as well
+    for cm := range channel {
+        consumer := cm.Consumer
+        msg := cm.Message
+        fmt.Printf("Consumer %s received a message, msgId: %v, content: '%s'\n",
+            consumer.Name(), msg.ID(), string(msg.Payload()))
+
+        consumer.Ack(msg)
+    }
+}
+```
+
+  </TabItem>
+</Tabs>
+````
+
+## Subscribe to topics
```

Comment:
It's better to move `Subscribe` and `Unsubscribe` sections before `Receive messages` sections.

Reply:
Done.